contracts/SunDAI.sol

```diff
@@ -0,0 +1,263 @@
+  IERC20 public dai;
+  mapping(address => uint256) public daiBalance;
+
+  constructor(address _daiAddr, address _bridgeAddr, bytes32 _name, bytes32 _symbol) public {
+    dai = IERC20(_daiAddr);
+    bridgeAddr = _bridgeAddr;
+    decimals = 18;
+    name = _name;
+    symbol = _symbol;
+  }
+
+  // 1. DAI is pulled into the contract, and record created
+  // 2. sunDAI are minted in amount of DAI deposited
+  // 3. sunDAI transfered to the Bridge
+  function _pullAndMint(address from, address to, uint256 value) internal {
+    require(dai.transferFrom(from, address(this), value));
+    daiBalance[from] = daiBalance[from].add(value);
+    _mint(to, value);
+  }
+
+  function _transferFrom(address from, address to, uint256 value) internal {
+    _transfer(from, to, value);
+
+    // only check allowance if other caller than bridge
+    if (to != bridgeAddr) {
+      approve(from, _allowed[from][msg.sender].sub(value));
+    }
+  }
+
+  // +------+         +--------+       +--------+
+  // |  DAI |         | sunDAI |       | Bridge |
+  // +---+--+         +----+---+       +----+---+
+  //     |                 |                |
+  //     |                 |                |
+  //     |                 | transferFrom   |
+  //     |  transferFrom   | <------------+ |
+  //     | <------------+  |                |
+  //     |                 | mint           |
+  //     | +------------>  | +--+           |
+  //     |                 |    |           |
+  //     |                 | <--+           |
+  //     |                 |                |
+  //     |                 | transfer       |
+  //     |                 | +----------->  |
+  //     |                 |                |
+  function transferFrom(address from, address to, uint256 value) public returns (bool) {
+    uint256 amountToMint = 0;
+    if (_balances[from] < value) {
+      amountToMint = value.sub(_balances[from]);
+      if (amountToMint > 0) {
+        require(msg.sender == bridgeAddr, "pullAndMint only for bridge");
+        require(to == bridgeAddr, "pullAndMint only to bridge");
+      }
+      _pullAndMint(from, msg.sender, amountToMint);
+    }
+    if (amountToMint < value) {
+      if (msg.sender == bridgeAddr || to == bridgeAddr) {
+        require(msg.sender == bridgeAddr, "transferFrom only by bridge to bridge");
+        require(to == bridgeAddr, "transferFrom only by bridge to bridge 2");
+      }
+      _transferFrom(from, to, value.sub(amountToMint));
+    }
+    return true;
+  }
+
+  function burn(address owner) public {
+    require(daiBalance[owner] > 0);
+    uint256 amount = _balances[owner];
+    if (daiBalance[owner] < amount) {
+      amount = daiBalance[owner];
+    }
+    daiBalance[owner] = daiBalance[owner].sub(amount);
+    _burn(owner, amount);
+    require(dai.transfer(owner, amount), "dai transfer failed");
+  }
+
+  function burnSender() public {
+    burn(msg.sender);
+  }
+
+
+
+  // default open zepplin code from here on
```

Comment:
Why do we pull this thing into here then? Can't we just inherit `ERC20` from openzeppelin?

Reply:
The balance mappings are private, so I cannot access if I inherit. :man_shrugging: am I stupid in some way?

Reply:
Ah ok. I guess using the methods instead is too inconvenient? 😄